src/scheduler/scheduler_test.cpp

```diff
@@ -0,0 +1,461 @@
+        *config_.get(), llm_engine_.get(), ssm_engine_.get());
+  }
+
+ private:
+  std::unique_ptr<SpeculativeScheduler> scheduler_;
+  std::unique_ptr<FakeLLMEngine> llm_engine_;
+  std::unique_ptr<FakeSSMEngine> ssm_engine_;
+  std::unique_ptr<SchedulerConfig> config_;
+  std::unique_ptr<Tokenizer> tokenizer_;
+};
+
+class RequestWrapper {
+ public:
+  RequestWrapper(Tokenizer* tokenizer)
+      : tokenizer_(tokenizer) {
+    sampling_param_.temperature = 0;
+    sampling_param_.top_p = 1.0;
+    sampling_param_.top_k = 0;
+    sampling_param_.repetition_penalty = 1.0;
+    sampling_param_.frequency_penalty = 0.0;
+    sampling_param_.presence_penalty = 0.0;
+
+    stopping_criteria_.max_tokens = 100;
+    stopping_criteria_.ignore_eos_token = false;
+    stopping_criteria_.eos_token_id = 2;
+  }
+
+  std::unique_ptr<Request> create_request(
+      const std::vector<int> prompt_tokens) {
+    return create_request_impl(prompt_tokens);
+  }
+
+  std::unique_ptr<Request> create_request(const std::string& prompt) {
+    std::vector<int> prompt_tokens;
+    tokenizer_->encode(prompt, &prompt_tokens);
+    return create_request_impl(prompt_tokens);
+  }
+
+  SequenceResult sequence_output() {
+    return outputs_["req_001"];
+  }
+
+ private:
+  std::unique_ptr<Request> create_request_impl(
+      const std::vector<int> prompt_tokens) {
+    const std::string request_id("req_001");
+    auto request = std::make_unique<Request>(request_id, prompt_tokens);
+    request->echo = true;
+    request->sampling_param = sampling_param_;
+    request->stopping_criteria = stopping_criteria_;
+    request->on_finish = [this, &request_id] (
+        const std::vector<SequenceResult>& seq_results,
+        const Status& status, const Statistics& stats) -> bool {
+      this->outputs_.emplace(request_id, seq_results[0]);
+      return true;
+    };
+    request->add_sequence([this, request_id] (
+        const std::string& delta, FinishReason reason) -> bool {
+      SequenceResult output;
+      output.output_text = delta;
+      output.finish_reason = reason;
+      this->outputs_.emplace(request_id, output);
+      return true;
+    });
+    return request;
+  }
+
+ private:
+  SamplingParameter sampling_param_;
+  StoppingCriteria stopping_criteria_;
+  Tokenizer* tokenizer_;
+
+  std::map<std::string, SequenceResult> outputs_;
+  bool is_finished = false;
+};
+
+TEST(SpeculativeSchedulerTest, Speculative4StepsPartiallyMatchTest) {
+  // who[1058] is[338] messi[4473] ?[29973]
+  const std::vector<int64_t> spec_token_ids = {1058, 338, 4473, 29973};
+  const std::vector<int64_t> valid_token_ids = {1058, 338, 4473, 1058, 338, 4473};
+  const uint64_t spec_steps = 4;
+
+  TestableSpeculativeScheduler scheduler(spec_token_ids, valid_token_ids,
+      spec_steps);
+  RequestWrapper request_wrapper(scheduler.tokenizer());
+
+  std::vector<int> input_tokens = {1058};
+  // expect output tokens: [1058] + [1058, 338, 4473, 1058] 
```

Comment:
why output has 2 who[1058] ?

Reply:
input 1058,  output is 1058,338,4473,1058

then, for the first stream output could be: 1058, 1058,338,4473,1058